__tests__/recursive.js

```diff
@@ -0,0 +1,184 @@
+                recursiveTransform: recursivelyTrimStrings,
+                baseState: {
+                    x: "  hello ",
+                    arr: [1, 2, 3]
+                },
+                desiredState: {
+                    x: "hello",
+                    arr: [1, 2, 3]
+                }
+            },
+            {
+                title:
+                    "when there is an unchanged leaf object in an unchanged array",
+                recursiveTransform: recursivelyTrimStrings,
+                baseState: {
+                    arr: [
+                        "!!!",
+                        {
+                            x: 42,
+                            y: "test"
+                        }
+                    ]
+                },
+                desiredState: {
+                    arr: [
+                        "!!!",
+                        {
+                            x: 42,
+                            y: "test"
+                        }
+                    ]
+                }
+            },
+            {
+                title:
+                    "when there is an unchanged leaf object in a changed array",
+                recursiveTransform: recursivelyTrimStrings,
+                baseState: {
+                    arr: [
+                        " !!!",
+                        {
+                            x: 42,
+                            y: "test"
+                        }
+                    ]
+                },
+                desiredState: {
+                    arr: [
+                        "!!!",
+                        {
+                            x: 42,
+                            y: "test"
+                        }
+                    ]
+                }
+            },
+            {
+                title:
+                    "when there is an unchanged leaf array in a changed object",
+                recursiveTransform: recursivelyTrimStrings,
+                baseState: {
+                    test: "!!! ",
+                    arr: [1, 2, 3]
+                },
+                desiredState: {
+                    test: "!!!",
+                    arr: [1, 2, 3]
+                }
+            },
+            {
+                title: "when there non-plain objects",
+                recursiveTransform: recursivelyTrimStrings,
+                baseState: {
+                    a: [nonPlainObject, 1],
+                    b: nonPlainObject
+                }
+            }
+        ]
+
+        testCases.forEach(
+            ({title, baseState, desiredState, recursiveTransform}) => {
+                it(`should not leave drafts ${title}`, () => {
+                    const next = recursiveTransform(baseState)
+                    if (desiredState) {
+                        // an exception is thrown during deep comparison if any of the drafts is not finalized
+                        expect(next).toEqual(desiredState)
```

Comment:
It might also good to add ` expect(next).not.toBe(baseState)` here, but this does not change the idea and the result of the test.